src/scanpy/metrics/_common.py

```diff
@@ -1,23 +1,85 @@
 from __future__ import annotations
 
 import warnings
+from abc import ABC, abstractmethod
+from dataclasses import InitVar, dataclass, field
 from functools import singledispatch
-from typing import TYPE_CHECKING, overload
+from typing import TYPE_CHECKING, ClassVar, overload
 
 import numpy as np
 import pandas as pd
 from scipy import sparse
 
-from .._compat import DaskArray
+from .._compat import DaskArray, fullname
 
 if TYPE_CHECKING:
     from typing import NoReturn, TypeVar
 
+    from anndata import AnnData
     from numpy.typing import NDArray
 
     T_NonSparse = TypeVar("T_NonSparse", bound=NDArray | DaskArray)
     V = TypeVar("V", bound=NDArray | sparse.csr_matrix)
 
+    _Vals = NDArray | sparse.spmatrix | DaskArray | pd.DataFrame | pd.Series
+
+
+__all__ = ["_get_graph", "_SparseMetric"]
+
+
+@dataclass
+class _SparseMetric(ABC):
+    g: sparse.csr_matrix
```

Comment:
I think we should use a name like `graph` or `weights` but I can be verbose sometimes